### DISCUSSION TITLE
Challenge

### DISCUSSION THREAD

**Discussion by sgamesdev** [Leo-Corporation/Gavilya#294] [Show and tell]
I challange you the deobsfucate your own dll 
[Gavilya.zip](https://github.com/Leo-Corporation/Gavilya/files/9639333/Gavilya.zip)

**Comment by sgamesdev** [comment]
Is my new obsufcation thechnology

**Comment by lpeyr** [comment]
Indeed, most methods aren't available, and there's a vast number of messy declarations with non-human readable names. However, all the classes and namespace names could still be retrieved by dotPeek. From this point on, it's easy to where the API keys code file is. Since I've written the code, I managed to identify the name of field where the key is stored as `byte[]`, despite the fact the dotPeek could decompile the method properly:
![image](https://user-images.githubusercontent.com/55453279/192114193-1e7e0433-386e-4018-bd52-fe2e2486a8f7.png)
Your obfuscator created useless messy methods that takes useless arguments since it only returns null values.
The following lines is where the key is decrypted:
~~~ cs
APIKeys.RAWGAPIKey = MxWLfii7QVobq7HMoUD.BpHZ0rOpM(J6aAFBiHcpmZKU9Pu3I.BpHZ0rOpM((string) APIKeys.HX9CGwoz6mAN11aPq4O(19120), t5AWBI0a5bP5Po16oVZ.BpHZ0rOpM((object) AhEZ0h001jJcMXoxQG6.BpHZ0rOpM(AhEZ0h001jJcMXoxQG6.XDc03bDqRT), APIKeys.YmaWxOhP3o, t5AWBI0a5bP5Po16oVZ.i0S0HagUfr), J6aAFBiHcpmZKU9Pu3I.BSG04jKQqj), (string) APIKeys.HX9CGwoz6mAN11aPq4O(19340), "", MxWLfii7QVobq7HMoUD.M1t07mQbBl);
~~~
We can see it reference the encryption key:
~~~ cs
APIKeys.YmaWxOhP3o
~~~
Don't know if it has anything related with API Keys, but found this (in `N4E7bnNTkkbXe712mK`):
~~~ cs
internal static SymmetricAlgorithm c9mngB8dE4()
{
      if (N4E7bnNTkkbXe712mK.uOMALAniVc())
        return (SymmetricAlgorithm) new AesCryptoServiceProvider();
      try
      {
        return (SymmetricAlgorithm) new RijndaelManaged();
      }
      catch
      {
        return (SymmetricAlgorithm) new AesCryptoServiceProvider();
      }
}

private static void sorAlTCEY7()
{
      try
      {
        RSACryptoServiceProvider.UseMachineKeyStore = true;
      }
      catch
      {
      }
}
~~~

----

I've also noticed that all methods in `Global` are returning either null values or default values (for `bool` notably). The code is hidden, probably in the other weird new namespaces.

----

To conclude, your obfuscation method is very efficient, however, it has some downsides, and therefore I won't use with Gavilya for the following reasons:
- It will affect performance; a lot of useless code is involved, which will slow down the app. (Downside with obfuscation)
- The sensitive part, which is about the API key, is protected efficiently, but the key is passed as plain text to the RAWG endpoint. If you install some network tools, I'm quite sure you will be able to intercept to request sent to the server, and thus retrieving the API key with ease.
- I think personally that obfuscating an Open-Source project doesn't really make any sense, all the code is available publicly.

But it perfectly fits for non-Open-Sourced projects: 

- Most of the methods are hidden, can't be decompiled
- The code is very hard to understand
- The readable part doesn't really make any sense and is - in most parts - useless to the de-obfuscation process.
- If you sell the product, it will take a lot of times for smart ones to crack it, making a painful journey.

Overall, I think you've made a great job; this a very performant, efficient obfuscator. Thanks for sharing this example with me 💯

**Comment by sgamesdev** [comment]
Thank you , yes it makes the app slower by 4%  , i suited this program for my app in order to hide the multiple encryption methods for my database credentials . Yes you cannot do anything with the api key , to fully protect it , With my database i encrypted the https request and the key is in my app and on the server so if the https request is intercepted , it's impossible to decrypt it without the encryption method and the key . The amount of the useless code can be adjusted from 1-9 .

**Comment by sgamesdev** [comment]
This is the only efficient way to protect your code

**Comment by sgamesdev** [comment]
Try to decompile 
[Gavilya.Fps.zip](https://github.com/Leo-Corporation/Gavilya/files/9639484/Gavilya.Fps.zip)

**Reply by lpeyr** [reply]
I just can't decompile it. dotPeek says it's "unsupported".

**Reply by sgamesdev** [reply]
Correct

**Reply by sgamesdev** [reply]
My program generates a native x86 EXE file stub . This way it is not possible to directly open in a decompiler . The decompiler decognizes  the app as a native EXE file .

**Reply by lpeyr** [reply]
Yes, I only have headers and Win32 ressources.

**Reply by sgamesdev** [reply]
Yes

**Reply by sgamesdev** [reply]
That's how steam is protected and many other big apps
